clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceRmdTTLFilter.java

```diff
@@ -27,68 +32,121 @@
-  protected final Map<RmdVersionId, Schema> rmdMapping;
+  private final long filterTimestamp;
+  private final HDFSSchemaSource schemaSource;
+  protected final Map<RmdVersionId, Schema> rmdSchemaMap;
+  protected final Map<Integer, Schema> valueSchemaMap;
   private final Map<RmdVersionId, RecordDeserializer<GenericRecord>> rmdDeserializerCache;
+  private final Map<Integer, RecordDeserializer<GenericRecord>> valueDeserializerCache;
+  private final Map<RmdVersionId, RecordSerializer<GenericRecord>> rmdSerializerCache;
+  private final Map<Integer, RecordSerializer<GenericRecord>> valueSerializerCache;
+  private final MergeRecordHelper mergeRecordHelper = new CollectionTimestampMergeRecordHelper();
 
   public VeniceRmdTTLFilter(final VeniceProperties props) throws IOException {
-    super(props);
+    super();
     ttlPolicy = TTLResolutionPolicy.valueOf(props.getInt(VenicePushJob.REPUSH_TTL_POLICY));
-    ttlInMs = TimeUnit.SECONDS.toMillis(props.getLong(VenicePushJob.REPUSH_TTL_IN_SECONDS));
-    schemaSource = new HDFSRmdSchemaSource(props.getString(VenicePushJob.RMD_SCHEMA_DIR));
-    rmdMapping = schemaSource.fetchSchemas();
+    long ttlInMs = TimeUnit.SECONDS.toMillis(props.getLong(VenicePushJob.REPUSH_TTL_IN_SECONDS));
+    long ttlStartTimestamp = props.getLong(VenicePushJob.REPUSH_TTL_START_TIMESTAMP);
+    this.filterTimestamp = ttlStartTimestamp - ttlInMs - 1;
+    this.schemaSource = new HDFSSchemaSource(
+        props.getString(VenicePushJob.VALUE_SCHEMA_DIR),
+        props.getString(VenicePushJob.RMD_SCHEMA_DIR));
+    this.rmdSchemaMap = schemaSource.fetchRmdSchemas();
+    this.valueSchemaMap = schemaSource.fetchValueSchemas();
     this.rmdDeserializerCache = new VeniceConcurrentHashMap<>();
+    this.valueDeserializerCache = new VeniceConcurrentHashMap<>();
+    this.rmdSerializerCache = new VeniceConcurrentHashMap<>();
+    this.valueSerializerCache = new VeniceConcurrentHashMap<>();
   }
 
   @Override
-  public boolean apply(final INPUT_VALUE value) {
+  public boolean checkAndMaybeFilterValue(final INPUT_VALUE value) {
     if (skipRmdRecord(value)) {
       return false;
     }
-    Instant curTime = Instant.now();
-    switch (ttlPolicy) {
-      case RT_WRITE_ONLY:
-        Instant timestamp = Instant.ofEpochMilli(getTimeStampFromRmdRecord(value));
-        return ChronoUnit.MILLIS.between(timestamp, curTime) > ttlInMs;
-      default:
-        throw new UnsupportedOperationException(ttlPolicy + " policy is not supported.");
+    if (Objects.requireNonNull(ttlPolicy) == TTLResolutionPolicy.RT_WRITE_ONLY) {
+      return validateByTTLandMaybeUpdateValue(value);
     }
+    throw new UnsupportedOperationException(ttlPolicy + " policy is not supported.");
   }
 
   @Override
   public void close() {
     schemaSource.close();
   }
 
-  public long getTimeStampFromRmdRecord(final INPUT_VALUE value) {
+  boolean validateByTTLandMaybeUpdateValue(final INPUT_VALUE value) {
```

Comment:
nit: rename to something like filterByTTLandMaybeUpdateValue

Reply:
Sure, fixed. 
Just to note that: the update value means updating the input value container. I think about this in the original apply() API name change and decided not to add "update value" there, as for partially stale value, the API actually "filters out" the old updates, and keeps the fresh ones.